spec/validation/configure_spec.rb

```diff
@@ -34,244 +41,264 @@
-      expect(run_configure_validation(correct_hash)).to be_empty
+      expect(run_configure_validation(correct_hash)).to eq(correct_hash)
     end
 
     it 'passes without questions key' do
       correct_hash.delete(:questions)
-      expect(run_configure_validation(correct_hash)).to be_empty
-    end
-
-    it 'checks that deep merged hashes pass (tests the testing)' do
-      h = correct_hash.deep_merge(domain: {
-                                    check_string_question: {
-                                      question: 'Am I deep merged into the domain?',
-                                      default: 'I sure hope so',
-                                    },
-                                  })
-      expect(run_configure_validation(h)).to be_empty
+      expect(run_configure_validation(correct_hash)).to eq(correct_hash)
     end
   end
 
   context 'with general invalid inputs' do
     it 'fails with invalid top level keys' do
       h = correct_hash.deep_merge(invalid_key: true)
-      expect(run_configure_validation(h).keys).to eq([:valid_top_level_keys])
+      expect_configure_error(h, /invalid top level key/)
     end
 
-    it 'fails if question is not a hash' do
-      h = correct_hash.deep_merge(group: {
-                                    question: 'Am I missing my mid level question key?',
-                                    type: 'string',
-                                    default: 'Each field will now be interpreted as a separate question',
-                                  })
-      results = build_validator(h).validate.errors
-      expect(results.keys).to eq([:parameters])
-      expect(results[:parameters][0]).to eq('must be a hash')
+    it 'fails if sections are not an array' do
+      h = correct_hash.deep_merge(group: { key: 'I am not an array' })
+      expect_configure_error(h, /must be an array/)
     end
   end
 
   context 'with invalid question fields' do
-    it 'fails if unrecognized fields in a question' do
-      h = correct_hash.deep_merge(domain: {
-                                    invalid_field_question: {
-                                      question: 'Do I have have an unrecognized field?',
-                                      default: 'I do',
-                                      invalid_filed: true,
-                                    },
-                                  })
-      expect(run_configure_validation(h).keys).to eq([:valid_top_level_question_keys])
-    end
-
-    it 'fails if question is missing a title' do
-      h = correct_hash.deep_merge(domain: {
-                                    missing_title_question: {
-                                      default: 'I am missing my question!',
-                                    },
-                                  })
-      results = run_configure_validation(h)
-      expect(results[:parameters][:question][0]).to eq('is missing')
+    it 'fails if the question is missing or empty identifier' do
```

Comment:
Generally if you need to use 'or' or 'and' in your test description it probably means you're testing multiple things in one test, and if you can break it up you should do. Doing this will make your tests clearer and more obvious why one is failing, since it will only be testing one thing. In this case since they don't share any setup it should be straightforward to split the two assertions out into two tests.
 Hide outdated

Reply:
Fixed in: https://github.com/alces-software/metalware/pull/185/commits/571722574003bcda5991dde8768e97f85b9452fa